docs/troubleshoot/app-framework/appserver-error-codes.md

```diff
@@ -2939,303 +2936,353 @@ The following error message codes may appear on the app-server log. Use the foll
+### ZWED0114E
+
+ The server found no plugin implementing the specified default authentication type of _type_.
 
   **Reason:**
 
-  TODO
+  The value of `components.app-server.dataserviceAuthentication.defaultAuthentication` within the server configuration specified a type of authentication that some authentication plugin must implement in order for the server to run. Because no plugin that successfully loaded declared that it implemented this type, the server found no implementation and could not continue.
 
   **Action:**
 
-  TODO
+  Review if any plugin you have implements the given type. If the type is incorrect, revise the configuration to choose a type that does exist in your system. If the type is correct, check if you are missing a required plugin.
 
 
 
-### ZWED0115E":"Unable to retrieve storage object from cluster. This is probably due to a timeout.\nYou may change the default of '%s' ms by setting 'node.cluster.storageTimeout' within the config. %s
+### ZWED0115E
+
+ Unable to retrieve storage object from cluster. This is probably due to a timeout.
+ You may change the default of '_storageTimeout_' ms by setting 'node.cluster.storageTimeout' within the config.
 
   **Reason:**
 
-  TODO
+  The app-server was running in cluster mode and a service attempted to get content from the cluster storage but this failed. Because storage could not be read, its possible the service that requested the storage will have further errors.
 
   **Action:**
 
-  TODO
+  If there was a network disruption or performance issue, a timeout could have occurred. Review the rest of the logs to see if there are other messages to explain the failure. You can attempt to avoid timeout-related failures by editing the configuration parameter `components.app-server.node.cluster.storageTimeout`.
 
 
 
-### ZWED0145E":"Cannot load SAF keyring content outside of z/OS
+### ZWED0145E
+
+ Cannot load SAF keyring content outside of z/OS
 
   **Reason:**
 
-  TODO
+  The Zowe configuration of `zowe.certificate` or `components.app-server.node.https` specifies SAF keyrings as locations to find keystore and truststore data. SAF keyrings only exist on z/OS, and the server detected it was not running on z/OS so it cannot continue.
 
   **Action:**
 
-  TODO
+  Modify the configuration to use a different keystore type, or migrate the server to z/OS.
 
 
 
-### ZWED0146E":"SAF keyring data had no attribute "%s". Attributes=
+### ZWED0146E
+
+ SAF keyring data had no attribute "_attribute_". Attributes=_attributeKeys_
 
   **Reason:**
 
-  TODO
+  Within the list of _attributeKeys_, _attribute_ could not be found.
 
   **Action:**
 
-  TODO
+  Check the keystore configuration of the server such as in `zowe.certificate` or `components.app-server.node.https` to see if it is valid for Zowe. The SAF keyring Zowe was configured to use may be missing a key and certificate pair, or certificate authorities keychain.
```

Comment:
Could be good to link this: https://docs.zowe.org/stable/user-guide/configure-certificates-keyring/#create-a-self-signed-jceracfks-certificate

80% sure this is out of date, but we will update it

Reply:
maybe. im not totally comfortable with that doc. i expect user errors to result.